### PR TITLE
Update rosetta integration

### PR DESCRIPTION
**Pull request by miiu96**
Update rosetta-sdk to version `v.0.7.0`

### CODE REVIEW

File: cmd/proxy/config/config.toml [hunk 1]
```diff
@@ -26,7 +26,7 @@
    # Otherwise, there are chances that only one full history node from a shard will process the requests
    BalancedFullHistoryNodes = true
 
-   # FaucetValue represents the default value for a faucet transaction. If set to "0", the faucet feature will be disabled
+   # FaucetValue represents the default value for a faucet transaction. If set to "0", the faucet feature will be offline
```
Comment (bogdan-rosianu) [inline cmd/proxy/config/config.toml:29]: revert change

File: cmd/proxy/main.go [hunk 2]
```diff
@@ -509,7 +508,10 @@ func createVersionsRegistry(
 	if err != nil {
 		return nil, err
 	}
+
 	if !isRosettaModeEnabled {
+		htbProc.StartCacheUpdate()
```
Comment (bogdan-rosianu) [inline cmd/proxy/main.go:513]: extend the condition so the proxy isn't offline?

File: observer/disabledNodesProvider.go [hunk 3]
```diff
@@ -32,7 +32,7 @@ func (d *disabledNodesProvider) GetAllNodes() ([]*data.NodeData, error) {
 
 // ReloadNodes return the desired return message as an error
 func (d *disabledNodesProvider) ReloadNodes(_ data.NodeType) data.NodesReloadResponse {
-	return data.NodesReloadResponse{Description: "disabled nodes provider", Error: d.returnMessage}
+	return data.NodesReloadResponse{Description: "offline nodes provider", Error: d.returnMessage}
```
Comment (bogdan-rosianu) [inline observer/disabledNodesProvider.go:35]: revert change

File: process/database/disabledElasticSearchConnector.go [hunk 4]
```diff
@@ -6,20 +6,20 @@ import (
 	"github.com/ElrondNetwork/elrond-proxy-go/data"
 )
 
-var errDatabaseConnectionIsDisabled = errors.New("database connection is disabled")
+var errDatabaseConnectionIsDisabled = errors.New("database connection is offline")
```
Comment (bogdan-rosianu) [inline process/database/disabledElasticSearchConnector.go:9]: revert changes in this file

File: process/disabled/epochStartNotifier.go [hunk 5]
```diff
@@ -5,20 +5,20 @@ import (
 	"github.com/ElrondNetwork/elrond-go/data"
 )
 
-// EpochStartNotifier represents a disabled struct that implements the EpochStartNotifier interface
+// EpochStartNotifier represents a offline struct that implements the EpochStartNotifier interface
```
Comment (bogdan-rosianu) [inline process/disabled/epochStartNotifier.go:8]: revert changes in this file

File: process/factory/faucetProcessorFactory.go [hunk 6]
```diff
@@ -24,7 +24,7 @@ func CreateFaucetProcessor(
 	pemFileLocation string,
 ) (facade.FaucetProcessor, error) {
 	if defaultFaucetValue.Cmp(big.NewInt(0)) == 0 {
-		log.Info("faucet is disabled")
+		log.Info("faucet is offline")
```
Comment (bogdan-rosianu) [inline process/factory/faucetProcessorFactory.go:27]: revert change

File: rosetta/README.md [hunk 7]
```diff
@@ -63,6 +63,12 @@ In order to stop the Observing Squad, run the command:
 make stop
 ```
 
+## Run offline 
+In order to can run elrond-rosetta server in offline mode, run the command:
```
Comment (bogdan-rosianu) [inline rosetta/README.md:67]: add an empty line between `##Run offline` and `In order to`

File: rosetta/api.go [hunk 8]
```diff
@@ -18,7 +19,78 @@ import (
...
+
+	offlineService := offline.NewOfflineService()
+
+	accountAPIController := server.NewAccountAPIController(offlineService, asserterServer)
+	blockAPIController := server.NewBlockAPIController(offlineService, asserterServer)
+	mempoolAPIController := server.NewMempoolAPIController(offlineService, asserterServer)
+
+	elrondProvider, err := provider.NewOfflineElrondProvider(elrondFacade, cfg.ElrondNetworkConfig)
+	if err != nil {
+		log.Error("cannot create elrond provider", "err", err)
```
Comment (bogdan-rosianu) [inline rosetta/api.go:55]: `"cannot create offline elrond provider"` ?

File: rosetta/api.go [hunk 9]
```diff
@@ -18,7 +19,78 @@ import (
...
+
+	router := server.NewRouter(
+		networkAPIController,
+		accountAPIController,
+		blockAPIController,
+		mempoolAPIController,
+		constructionAPIController,
+	)
+
+	loggedRouter := server.LoggerMiddleware(router)
```
Comment (bogdan-rosianu) [inline rosetta/api.go:76]: loggerRouter*

File: rosetta/api.go [hunk 10]
```diff
@@ -18,7 +19,78 @@ import (
...
+		Addr:    fmt.Sprintf(":%d", port),
+		Handler: corsRouter,
+	}
+
+	log.Info("elrond rosetta server is in offline mode")
+
+	return httpServer, nil
+}
+
+func createServerOnline(
```
Comment (bogdan-rosianu) [inline rosetta/api.go:89]: createOnlineServer

File: rosetta/config.toml [hunk 11]
```diff
@@ -3,27 +3,37 @@
    # ServerPort is the port used for the web server. The frontend will connect to this port
    ServerPort = 8079
 
-   # RequestTimeoutSec represents the maximum number of seconds a request can last until throwing an error
-   RequestTimeoutSec = 10
+    # RequestTimeoutSec represents the maximum number of seconds a request can last until throwing an error
```
Comment (bogdan-rosianu) [inline rosetta/config.toml:6]: bad formatting. revert it please

File: rosetta/configuration/configuration.go [hunk 12]
```diff
@@ -16,21 +16,39 @@ const (
...
 	Peers                  []*types.Peer
 }
 
-//LoadConfiguration will load configuration
+// LoadConfiguration will load configuration
 func LoadConfiguration(networkConfig *provider.NetworkConfig, generalConfig *config.Config) *Configuration {
+	return loadConfig(networkConfig, generalConfig)
+}
+
+func LoadOfflineConfig(generalConfig *config.Config) *Configuration {
```
Comment (bogdan-rosianu) [inline rosetta/configuration/configuration.go:41]: add comment

File: rosetta/provider/offline_provider.go [hunk 13]
```diff
@@ -0,0 +1,18 @@
+package provider
+
+import (
+	"github.com/ElrondNetwork/elrond-proxy-go/api"
+)
+
+func NewOfflineElrondProvider(elrondFacade api.ElrondProxyHandler, networkConfig *NetworkConfig) (*ElrondProvider, error) {
```
Comment (bogdan-rosianu) [inline rosetta/provider/offline_provider.go:7]: comment

File: rosetta/services/offline/offline_service_all.go [hunk 14]
```diff
@@ -0,0 +1,81 @@
...
+import (
+	"context"
+
+	"github.com/ElrondNetwork/elrond-proxy-go/rosetta/services"
+	"github.com/coinbase/rosetta-sdk-go/types"
+)
+
+type offlineService struct{}
+
+func NewOfflineService() *offlineService {
```
Comment (bogdan-rosianu) [inline rosetta/services/offline/offline_service_all.go:12]: comment

**Review by bogdan-rosianu — CHANGES_REQUESTED** [review]

File: cmd/proxy/config/config.toml (hunk 1 above)
Comment (miiu96) [reply cmd/proxy/config/config.toml:29]: done.

File: observer/disabledNodesProvider.go (hunk 3 above)
Comment (miiu96) [reply observer/disabledNodesProvider.go:35]: done.

File: process/database/disabledElasticSearchConnector.go (hunk 4 above)
Comment (miiu96) [reply process/database/disabledElasticSearchConnector.go:9]: done.

File: process/disabled/epochStartNotifier.go (hunk 5 above)
Comment (miiu96) [reply process/disabled/epochStartNotifier.go:8]: done.

File: process/factory/faucetProcessorFactory.go (hunk 6 above)
Comment (miiu96) [reply process/factory/faucetProcessorFactory.go:27]: done.

File: rosetta/README.md (hunk 7 above)
Comment (miiu96) [reply rosetta/README.md:67]: done.

File: rosetta/api.go (hunk 8 above)
Comment (miiu96) [reply rosetta/api.go:55]: done.

File: rosetta/api.go (hunk 9 above)
Comment (miiu96) [reply rosetta/api.go:76]: done.

File: rosetta/api.go (hunk 10 above)
Comment (miiu96) [reply rosetta/api.go:89]: done.

File: rosetta/config.toml (hunk 11 above)
Comment (miiu96) [reply rosetta/config.toml:6]: done.

File: rosetta/configuration/configuration.go (hunk 12 above)
Comment (miiu96) [reply rosetta/configuration/configuration.go:41]: added.

File: rosetta/provider/offline_provider.go (hunk 13 above)
Comment (miiu96) [reply rosetta/provider/offline_provider.go:7]: added

File: rosetta/services/offline/offline_service_all.go (hunk 14 above)
Comment (miiu96) [reply rosetta/services/offline/offline_service_all.go:12]: added.

File: rosetta/api.go [hunk 15]
```diff
@@ -18,7 +19,78 @@ import (
...
+	elrondFacade api.ElrondProxyHandler,
+	generalConfig *config.Config,
+	port int,
+	isOffline bool,
+) (*http.Server, error) {
+	if !isOffline {
+		return createOnlineServer(elrondFacade, generalConfig, port)
+	}
+
+	cfg := configuration.LoadOfflineConfig(generalConfig)
```
Comment (AdoAdoAdo) [inline rosetta/api.go:32]: This method and createOnlineServer can be refactored to remove duplicated code.
some of the lines 32-> 53 can be extracted into a CreateProvider method that receives either an offline or an online service
some of the lines 59 -> 82 can be extracted into a CreateHttpServer method

File: rosetta/services/construction_service_test.go [hunk 16]
```diff
@@ -68,8 +68,8 @@ func TestConstructionAPIService_ConstructionPreprocess(t *testing.T) {
 	feeMultiplier := 1.1
 	maxFee := "1234567"
 
-	gasPrice := uint64(100)
-	gasLimit := uint64(10000)
+	gasPrice := float64(100)
```
Comment (AdoAdoAdo) [inline rosetta/services/construction_service_test.go:71]: why are these float?

File: rosetta/services/construction_service.go [hunk 17]
```diff
@@ -93,12 +96,12 @@ func (cas *constructionAPIService) checkOperationsAndMeta(ops []*types.Operation
 	}
 
 	if meta["gasLimit"] != nil {
-		if _, ok := meta["gasLimit"].(uint64); !ok {
+		if _, ok := meta["gasLimit"].(float64); !ok {
```
Comment (AdoAdoAdo) [inline rosetta/services/construction_service.go:99]: why are these float values?

**Review by AdoAdoAdo — CHANGES_REQUESTED** [review]

File: process/database/disabledElasticSearchConnector.go [hunk 18]
```diff
@@ -14,12 +14,12 @@ func NewDisabledElasticSearchConnector() *disabledElasticSearchConnector {
 	return new(disabledElasticSearchConnector)
 }
 
-// GetTransactionsByAddress will return error because database connection is disabled
+// GetTransactionsByAddress will return error because database connection is offline
```
[the remark this answers is not in the document]
Comment (miiu96) [reply process/database/disabledElasticSearchConnector.go:17]: done.

File: process/disabled/epochStartNotifier.go [hunk 19]
```diff
@@ -9,16 +9,16 @@ import (
 type EpochStartNotifier struct {
 }
 
-// RegisterNotifyHandler won't do anything as this is a disabled component
+// RegisterNotifyHandler won't do anything as this is a offline component
```
[the remark this answers is not in the document]
Comment (miiu96) [reply process/disabled/epochStartNotifier.go:12]: done.

File: rosetta/README.md [hunk 20]
```diff
@@ -63,6 +63,14 @@ In order to stop the Observing Squad, run the command:
 make stop
 ```
 
+## Run offline 
+
+In order to can run elrond-rosetta server in offline mode, run the command:
```
[the remark this answers is not in the document]
Comment (miiu96) [reply rosetta/README.md:68]: refactored.

File: rosetta/api.go (hunk 15 above)
Comment (miiu96) [reply rosetta/api.go:32]: refactored.

File: rosetta/services/construction_service_test.go (hunk 16 above)
Comment (miiu96) [reply rosetta/services/construction_service_test.go:71]: fixed

File: rosetta/services/construction_service.go (hunk 17 above)
Comment (miiu96) [reply rosetta/services/construction_service.go:99]: fixed

**Review by AdoAdoAdo — DISMISSED** [review]

**Review by bogdan-rosianu — DISMISSED** [review]

File: rosetta/offline_config_devnet.toml [hunk 21]
```diff
@@ -0,0 +1,5 @@
+# OfflineSettings section of the rosetta server offline mod
+[OfflineSettings]
+    ChainID = "D"
+    MinGasPrice = 50000
```
Comment (bogdan-rosianu) [inline rosetta/offline_config_devnet.toml:4]: should be switched (min gas limit 50k, min gas price 1000000000)

File: rosetta/offline_config_mainnet.toml [hunk 22]
```diff
@@ -0,0 +1,5 @@
+# OfflineSettings section of the rosetta server offline mod
+[OfflineSettings]
+    ChainID = "1"
+    MinGasPrice = 50000
+    MinGasLimit = 1000000000
```
Comment (bogdan-rosianu) [inline rosetta/offline_config_mainnet.toml:5]: should be switched (min gas limit 50k, min gas price 1000000000)
+ empty line

File: rosetta/offline_config_mainnet.toml [hunk 23]
```diff
@@ -0,0 +1,5 @@
+# OfflineSettings section of the rosetta server offline mod
```
Comment (bogdan-rosianu) [inline rosetta/offline_config_mainnet.toml:1]: mode*

File: rosetta/offline_config_devnet.toml [hunk 24]
```diff
@@ -0,0 +1,5 @@
+# OfflineSettings section of the rosetta server offline mod
```
Comment (bogdan-rosianu) [inline rosetta/offline_config_devnet.toml:1]: mode*

File: rosetta/rosetta-cli-config/mainnet/elrond_mainnet.json [hunk 25]
```diff
@@ -3,174 +3,27 @@
...
   }
+ },
+ "data": {
+  "historical_balance_enabled": true,
+  "reconciliation_disabled": true,
+  "inactive_discrepency_search_disabled": true,
+  "balance_tracking_disabled": true
  }
-}
+}
```
Comment (bogdan-rosianu) [inline rosetta/rosetta-cli-config/mainnet/elrond_mainnet.json:29]: empty line

File: rosetta/rosetta-cli-config/mainnet/elrond.ros [hunk 26]
```diff
@@ -0,0 +1,91 @@
...
+        "type":"Transfer",
+        "account":{{recipient.account_identifier}},
+        "amount":{
+          "value":{{recipient_amount}},
+          "currency":{{elrond_currency}}
+        }
+      }
+    ];
+  }
+}
```
Comment (bogdan-rosianu) [inline rosetta/rosetta-cli-config/mainnet/elrond.ros:91]: empty line

File: rosetta/rosetta.sh [hunk 27]
```diff
@@ -59,36 +66,37 @@ echo -e
...
 done
 
 #Say what has been started
-echo "testnet" > ~/.squadlocation
+echo "devnet" > ~/.squadlocation
 
 #Start the stack
 echo -e
-echo -e "${GREEN}--> Starting the Observer+Proxy TestNet Stack...${NC}"
+echo -e "${GREEN}--> Starting the Observer+Proxy DEVNET Stack...${NC}"
```
Comment (bogdan-rosianu) [inline rosetta/rosetta.sh:99]: should we keep this for testnet as well?

**Review by bogdan-rosianu — CHANGES_REQUESTED** [review]

File: rosetta/offline_config_devnet.toml (hunk 21 above)
Comment (miiu96) [reply rosetta/offline_config_devnet.toml:4]: done.

File: rosetta/offline_config_mainnet.toml (hunk 22 above)
Comment (miiu96) [reply rosetta/offline_config_mainnet.toml:5]: done.

File: rosetta/offline_config_mainnet.toml (hunk 23 above)
Comment (miiu96) [reply rosetta/offline_config_mainnet.toml:1]: done.

File: rosetta/offline_config_devnet.toml (hunk 24 above)
Comment (miiu96) [reply rosetta/offline_config_devnet.toml:1]: done.

File: rosetta/rosetta-cli-config/mainnet/elrond_mainnet.json (hunk 25 above)
Comment (miiu96) [reply rosetta/rosetta-cli-config/mainnet/elrond_mainnet.json:29]: added

File: rosetta/rosetta-cli-config/mainnet/elrond.ros (hunk 26 above)
Comment (miiu96) [reply rosetta/rosetta-cli-config/mainnet/elrond.ros:91]: added

File: rosetta/rosetta.sh (hunk 27 above)
Comment (miiu96) [reply rosetta/rosetta.sh:99]: no. just for devnet and mainnet because testnet is reset more ofen

File: rosetta/offline_config_mainnet.toml [hunk 28]
```diff
@@ -1,5 +1,5 @@
-# OfflineSettings section of the rosetta server offline mod
+# OfflineSettings section of the rosetta server offline mode
 [OfflineSettings]
     ChainID = "1"
-    MinGasPrice = 50000
-    MinGasLimit = 1000000000
\ No newline at end of file
+    MinGasPrice = 1000000000
+    MinGasLimit = 50000
```
Comment (bogdan-rosianu) [inline rosetta/offline_config_mainnet.toml:5]: empty line

**Review by bogdan-rosianu — CHANGES_REQUESTED** [review]

File: rosetta/offline_config_mainnet.toml (hunk 28 above)
Comment (miiu96) [reply rosetta/offline_config_mainnet.toml]: done./

File: rosetta/rosetta-cli-config/devnet/elrond_devnet.json [hunk 29]
```diff
@@ -0,0 +1,30 @@
...
+    }
+  },
+  "data": {
+    "start_index": 2216150,
+    "historical_balance_enabled": true,
+    "reconciliation_disabled": true,
+    "inactive_discrepency_search_disabled": true,
+    "balance_tracking_disabled": true
+  }
+}
```
Comment (AdoAdoAdo) [inline rosetta/rosetta-cli-config/devnet/elrond_devnet.json:30]: newline at eof

**Review by bogdan-rosianu — DISMISSED** [review]

**Review by AdoAdoAdo — DISMISSED** [review]

File: rosetta/rosetta-cli-config/mainnet/elrond_mainnet.json [hunk 30]
```diff
@@ -3,174 +3,27 @@
   "blockchain": "Elrond",
   "network": "1"
  },
- "online_url": "http://10.0.0.2:8079",
+ "offline_url": "http://localhost:8080",
```
[the remark this answers is not in the document]
Comment (miiu96) [reply rosetta/rosetta-cli-config/mainnet/elrond_mainnet.json]: changed.

File: rosetta/rosetta-cli-config/devnet/elrond_devnet.json (hunk 29 above)
Comment (miiu96) [reply rosetta/rosetta-cli-config/devnet/elrond_devnet.json:30]: added

File: rosetta/api.go [hunk 31]
```diff
@@ -18,7 +19,58 @@ import (
...
+
+	offlineService := offline.NewOfflineService()
+
+	accountAPIController := server.NewAccountAPIController(offlineService, asserterServer)
+	blockAPIController := server.NewBlockAPIController(offlineService, asserterServer)
+	mempoolAPIController := server.NewMempoolAPIController(offlineService, asserterServer)
+
+	elrondProvider, err := provider.NewOfflineElrondProvider(elrondFacade, cfg.ElrondNetworkConfig)
+	if err != nil {
+		log.Error("cannot create offline elrond provider", "err", err)
```
[the remark this answers is not in the document]
Comment (miiu96) [reply rosetta/api.go:51]: removed.

File: rosetta/api.go [hunk 32]
```diff
@@ -102,3 +145,24 @@ func CreateServer(elrondFacade api.ElrondProxyHandler, generalConfig *config.Con
...
+		false,
+		[]*types.NetworkIdentifier{
+			network,
+		},
+		nil,
+		false,
+		"",
+	)
+	if err != nil {
+		log.Error("cannot create asserter", "err", err)
```
[the remark this answers is not in the document]
Comment (miiu96) [reply rosetta/api.go:163]: removed.

File: rosetta/provider/common.go [hunk 33]
```diff
@@ -15,6 +15,7 @@ type NetworkConfig struct {
 	MinTxVersion   uint32 `json:"erd_min_transaction_version"`
 	StartTime      uint64 `json:"erd_start_time"`
 	RoundDuration  uint64 `json:"erd_round_duration"`
+	NodeIsStarting string `json:"error"`
```
[the remark this answers is not in the document]
Comment (miiu96) [reply rosetta/provider/common.go:18]: This field is the field from the API

File: rosetta/provider/elrond_provider.go [hunk 34]
```diff
@@ -58,6 +59,7 @@ func (ep *ElrondProvider) initializeElrondProvider() error {
 	for count := 0; count < MaxRetriesGetNetworkConfig; count++ {
 		networkConfig, err = ep.GetNetworkConfig()
 		if err != nil {
+			log.Info("will try again", "count", count)
```
[the remark this answers is not in the document]
Comment (miiu96) [reply rosetta/provider/elrond_provider.go:62]: done.

File: rosetta/provider/elrond_provider.go [hunk 35]
```diff
@@ -110,8 +117,18 @@ func (ep *ElrondProvider) GetNetworkConfig() (*NetworkConfig, error) {
 	}
 
 	err = json.Unmarshal(responseBytes, networkConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	if networkConfig.NodeIsStarting != "" {
+		err = errors.New(networkConfig.NodeIsStarting)
+		log.Warn("cannot get network metrics", "error", err.Error())
```
[the remark this answers is not in the document]
Comment (miiu96) [reply rosetta/provider/elrond_provider.go:126]: removed.

File: rosetta/provider/offline_provider.go [hunk 36]
```diff
@@ -0,0 +1,19 @@
+package provider
+
+import (
+	"github.com/ElrondNetwork/elrond-proxy-go/api"
+)
+
+// NewOfflineElrondProvider will create an elrond provided in offline mode
```
[the remark this answers is not in the document]
Comment (miiu96) [reply rosetta/provider/offline_provider.go:7]: done.

File: rosetta/rosetta-cli-config/devnet/elrond.ros [hunk 37]
```diff
@@ -0,0 +1,91 @@
...
+        "type":"Transfer",
+        "account":{{recipient.account_identifier}},
+        "amount":{
+          "value":{{recipient_amount}},
+          "currency":{{elrond_currency}}
+        }
+      }
+    ];
+  }
+}
```
[the remark this answers is not in the document]
Comment (miiu96) [reply rosetta/rosetta-cli-config/devnet/elrond.ros:91]: added

File: rosetta/rosetta-cli-config/devnet/elrond_devnet.json [hunk 38]
```diff
@@ -0,0 +1,30 @@
+{
+  "network": {
+    "blockchain": "Elrond",
+    "network": "D"
+  },
+  "online_url": "http://localhost:8080",
+  "data_directory": "",
+  "http_timeout": 200,
+  "max_retries": 20,
+  "tip_delay": 100,
```
[the remark this answers is not in the document]
Comment (miiu96) [reply rosetta/rosetta-cli-config/devnet/elrond_devnet.json:10]: number of seconds to wait before starting the scenario, changed to 10

**Review by SebastianMarian — APPROVED** [review]

**Review by bogdan-rosianu — APPROVED** [review]

**Review by AdoAdoAdo — APPROVED** [review]

**Pull request merged**